pxr/usdImaging/usdImaging/delegate.cpp

```diff
@@ -282,11 +282,17 @@ UsdImagingDelegate::_GetHdPrimInfo(const SdfPath &cachePath)
 }
 
 Usd_PrimFlagsConjunction 
-UsdImagingDelegate::_GetDisplayPredicate() const
+UsdImagingDelegate::_GetDisplayPredicate(bool forPrototypes) const
 {
-    return _displayUnloadedPrimsWithBounds ?
-            UsdPrimIsActive && UsdPrimIsDefined && !UsdPrimIsAbstract :
-            UsdPrimDefaultPredicate;
+    Usd_PrimFlagsConjunction retval = UsdPrimIsActive && !UsdPrimIsAbstract;
+    if (!forPrototypes) {
+        retval = retval && UsdPrimIsDefined;
```

Comment:
[nested_ptinst.zip](https://github.com/PixarAnimationStudios/USD/files/5944002/nested_ptinst.zip)

Thanks Spiff-- here's a zip that has a point instancer with a prototype in an over, and a def underneath, in sphereInstancer.usda -- it doesn't draw as expected unless I remove the UsdPrimHasDefiningSpecifier (or 
UsdPrimIsDefined) in the predicate. Am I misunderstanding something?

There's also an ni_pi.usda in the zip that puts the point instancer behind scenegraph instancing. No predicate as yet seems to get that drawing as expected. I'll continue investigating, unless you already know what's going on.

Thanks again!

Reply:
Hmm.. the reason I thought this might not be working doesn't hold up.  There's alot going on right now, but I'll try to take a deeper look and get back to you.  First thing to check is whether, outside of imaging, the prototype root prims (which presumably are children of an undefined "over") pass the predicate.  If they do not, then maybe we have a core bug.

Reply:
Hi Spiff, in the sphereInstancer.usda within the zip archive I posted above, I ran the following in usdview.

I think that is correct from a core standpoint, so I don't think there's a way to get the prototype detection logic to descend into overs without removing Usd.PrimHasDefiningSpecifier . But I think I understand your point that you want the prim range to descend into overs, but do not honor prototypes that themselves are pure overs. I'd have to think more about the right logic for that, if that's what the standard should be.

(I haven't had time to dig more into the nested instancing case yet in the zip archive).

Thanks!

(Dummy is an over, Dummy/sphere is a def)

>>> [x for x in Usd.PrimRange( usdviewApi.stage.GetPrimAtPath('/World/Instancer'), Usd.PrimIsActive & Usd.PrimHasDefiningSpecifier & ~Usd.PrimIsAbstract & Usd.PrimIsLoaded)]
[Usd.Prim(</World/Instancer>)]
>>> 
>>> [x for x in Usd.PrimRange( usdviewApi.stage.GetPrimAtPath('/World/Instancer'), Usd.PrimIsActive & ~Usd.PrimIsAbstract & Usd.PrimIsLoaded)]
[Usd.Prim(</World/Instancer>), Usd.Prim(</World/Instancer/Dummy>), Usd.Prim(</World/Instancer/Dummy/sphere>)]

Reply:
What we were thinking, and what I thought the example showed, is that the parent of the prototypes (itself a child of the PI) would be a pure over. In our experience, having the prototypes themselves be overs would not even be practical to achieve, since they are references to defined models.

Reply:
That's correct-- in my attached scene, Dummy is a pure over beneath the PI, and Sphere is a def for the prototype under Dummy. But with UsdPrimHasDefiningSpecifier, traversals won't reach Dummy from the PI.

Reply:
But it shouldn't need to traverse from the PI, by my read of the code: https://github.com/PixarAnimationStudios/USD/blob/67af68bac3469b328e7f9dff9e0dc6e198ea7a5f/pxr/usdImaging/usdImaging/pointInstancerAdapter.cpp#L199-L233 . It explicitly starts from each prototype root, gleaned via the _prototypes_ relationship.  I'm not super familiar with this code, though.

Reply:
Good point, Spiff. I reran the snippet above on the prototype root, which is a pure over with Sphere def'd under it.

>>> [x for x in Usd.PrimRange( usdviewApi.stage.GetPrimAtPath('/World/Instancer/Dummy'), Usd.PrimIsLoaded & ~Usd.PrimIsAbstract & Usd.PrimIsActive)]
[Usd.Prim(</World/Instancer/Dummy>), Usd.Prim(</World/Instancer/Dummy/sphere>)]

>>> [x for x in Usd.PrimRange( usdviewApi.stage.GetPrimAtPath('/World/Instancer/Dummy'), Usd.PrimIsLoaded & ~Usd.PrimIsAbstract & Usd.PrimIsActive & Usd.PrimHasDefiningSpecifier)]
[]

I think the predicate gets applied right away, which makes sense, so it won't descend below the pure over of Dummy.

I kept UsdPrimHasDefiningSpecifier out of the predicate, but now have it as a condition alongside checking for the adapter, such that pure overs are skipped, but defs that have ancestral overs are honored.

I'm still debugging the nested instancing case.

Thanks again!

Reply:
Aaron, I thought we’d agreed above yhat the proto roots themselves are not meant to be pure overs, but their parents can be... a “protos” Xform or scope, if you will. Can you keep to that design? Then we can have nice simple code.

Reply:
Sorry Spiff--- I've been scatterbrained. Fixed. Will continue investigating nested instances and regression testing.